tests/StatsdClient.Tests/DogStatsdServiceTests.cs

```diff
@@ -15,33 +15,18 @@ private List<string> ReceiveData(IDogStatsd dogStatsdInstance, string testServer
         {
             using (var udpListener = new UdpListener(testServerName, testPort))
             {
-                var listenThread = new Thread(udpListener.Listen);
+                var listenThread = new Thread(udpListener.ListenAndWait);
                 listenThread.Start();
 
                 sendData();
 
-                while (listenThread.IsAlive) ;
-
+                udpListener.Shutdown();
```

Comment:
It was working before because the timing test was relying on something happening in less that 100ms.

In the test default_port_is_8125() because I change the timeout to 1 second, now the thread running the UDP listener has a larger timeout before it actually finishes and publishes the metric, meanwhile the test thread races on and hits the assert before the metric has been issued.

You could solve this in a horrible way by having the test thread also sleep for a longer amount of time, but a much easier way is to design the UDP listener to block but not exit until explicitly called, this way it is deterministic.

This is also because the blocking actually has a timeout and what also happens is the blocking call timesout BEFORE the thread sleep for the metric test and then the listener exists and misses the metric from the timing loop.

Easiest way is to checkout master and modify the thread.sleep in default_port_is_8125() and you will see what I mean there.

Also, the listener only has a very small chance of exiting without receiving the data, the race condition is around the blocking call timing out at the same time the data is sent (and thus missing it) but the shutdown setting the boolean and causing the loop to exit.

This could be fixed by also telling the udp listener to wait for N messages if you know in advance how many messages are being sent.